### PR TITLE
Discrete choice

### PR DESCRIPTION
**Pull request by NathanielF**
# Discrete Choice Modelling

Working out some of the details of how to represent the various incarnations of the discrete choice models in PyMC building to the hierarchical or random coefficients logit described for instance in Jim Savage's work in Stan here: https://khakieconomics.github.io/2019/03/17/Logit-models-of-discrete-choice.html. They are quite distinct from reinforcement learning strategies and are used to estimate aggregate demand and supply curves in differentiated goods markets and more individually looking at the substitution pattern customers have between goods. 

This is a draft primarily because i need to find an efficient representation of the models in PyMC and write up more detail on motivating the models. 

+ [x] Notebook follows style guide https://docs.pymc.io/en/latest/contributing/jupyter_style.html
+ [x] PR description contains a link to the relevant issue:
  * a tracker one for existing notebooks (tracker issues have the "tracker id" label)
  * or a proposal one for new notebooks
+ [x] Check the notebook is not excluded from any pre-commit check: https://github.com/pymc-devs/pymc-examples/blob/main/.pre-commit-config.yaml

### Helpful links
* https://github.com/pymc-devs/pymc-examples/blob/main/CONTRIBUTING.md

### CODE REVIEW

File: examples/generalized_linear_models/GLM-discrete-choice_models.ipynb [hunk 1]
```diff
@@ -0,0 +1,4369 @@
+{
```
[the remark this answers is not in the document]
Comment (NathanielF) [reply examples/generalized_linear_models/GLM-discrete-choice_models.ipynb:1]: Hi @twiecki 

To be honest. I think this one needs quite a bit more work. The modelling is brittle to different randomly generated data and I'm not sure I have the right approach in general. 

Experimenting with it when I can. I think this is an interesting class of models, but getting the translation from Stan right is proving a bit trickier than I hoped.

File: examples/generalized_linear_models/GLM-discrete-choice_models.ipynb [hunk 2]
```diff
@@ -0,0 +1,5710 @@
...
```
Comment (ricardoV94) [inline examples/generalized_linear_models/GLM-discrete-choice_models.ipynb:1]: <p>You can use <code>pm.math.softmax</code> to avoid that annoying warning </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/pymc-devs/pymc-examples/pull/544/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='4'/>

Comment (NathanielF) [reply examples/generalized_linear_models/GLM-discrete-choice_models.ipynb:1]: Thanks, yeah I'll adjust that. I think the modelling strategy is coming along now. I got overly hung up on the long-data format stan was using. Sticking with the wide data format for the minute. Seems more transparent to me and better pedagogically.

**Comment by NathanielF** [comment]
Marking this one as ready for review @twiecki and @ricardoV94. 

I've found a representation of the discrete choice models that i'm pretty happy with. I'm demonstrating them on two different but canonical data sets. 

(1) On the choice of heating systems, where i focus just on specifying the utility matrix for the individual alternatives
(2) On the choice of crackers with repeated decisions over the same decision maker. 

In (1) i demonstrate how you can add alternative specific parameters e.g. intercepts and beta parameters for income on the specific alternative. 

In (2) i focus on how you can add person specific modifications of utility and how you can use prior constraints in the Bayesian context to ensure that the parameter estimates "make sense" i.e. negative parameter estimates for the effect of price on utility. 

All models fit well, and in reasonable time. However, in (2) i've truncated the data set a little because i ran into a bracket nesting error on the full data set. Would keen to know how to replace my for-loop here with scan.... but i wasn't sure how to do that...,.?

I think i'll likely add some more to the text-write up, but would like some interim feedback if you have any on the modelling design.

@drbenvincent in case of interest.

**Comment by NathanielF** [comment]
<p>But i do use the probabilities 'p' in nearly every plot, It's kind of one of the main quantities of interest. I didn't wrap the utilities for the reason you mention...</p>

---

 <a href='https://app.reviewnb.com/pymc-devs/pymc-examples/pull/544/discussion/#comment-1597516526'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

**Comment by NathanielF** [comment]
<p>Yep. Will adjust next commit. </p>

---

 <a href='https://app.reviewnb.com/pymc-devs/pymc-examples/pull/544/discussion/#comment-1597516540'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

**Comment by NathanielF** [comment]
<p>It's possibly not needed. I wrote it in the way which was most intuitive to me. I'd be keen to understand how a fancy index approach could work? </p><p> </p><p>I don't think we have any non-square matrices that we're looking to construct. But we do have varying length sets of rows per person. So we might have something like: </p><p> </p><p>| Person ID | Choice ID | Choice&nbsp;|</p><p>|-----------|-----------|---------|</p><p>| Person 1&nbsp;| 1&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;| Nabisco |</p><p>| Person 1&nbsp;| 2&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;| Keebler |</p><p>| Person 2&nbsp;| 1&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;| Nabisco |</p><p>| Person 2&nbsp;| 2&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;| Nabisco |</p><p>| Person 2&nbsp;| 3&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;| Nabisco |</p><p> </p><p>So i'm building per person 3 equations and i want to add an alternative specific beta-coefficient to the alternative specific intercept alpha. In this case i'd need the person specific beta coefficient to be applied to the first two rows and then the next three...  It was just allot of indexes to keep in my head, but if you have ideas about how to make it cleaner, that'd be great!?</p>

---

 <a href='https://app.reviewnb.com/pymc-devs/pymc-examples/pull/544/discussion/#comment-1597516556'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

**Comment by NathanielF** [comment]
<p>Oh actually i think i see it now! </p>

---

 <a href='https://app.reviewnb.com/pymc-devs/pymc-examples/pull/544/discussion/#comment-1597516556'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

**Comment by NathanielF** [comment]
<p>Yeah, ok I think i have it now. </p>

---

 <a href='https://app.reviewnb.com/pymc-devs/pymc-examples/pull/544/discussion/#comment-1597516556'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

**Comment by NathanielF** [comment]
<p>Resolved</p>

---

 <a href='https://app.reviewnb.com/pymc-devs/pymc-examples/pull/544/discussion/#comment-1597516540'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

**Comment by NathanielF** [comment]
<p>Updated in latest commit. Thanks for the push. This is much neater. </p>

---

 <a href='https://app.reviewnb.com/pymc-devs/pymc-examples/pull/544/discussion/#comment-1597516556'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

**Comment by NathanielF** [comment]
<p>Removed any redundant Deterministics. </p>

---

 <a href='https://app.reviewnb.com/pymc-devs/pymc-examples/pull/544/discussion/#comment-1597516526'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

**Comment by NathanielF** [comment]
Incorporated all those changes now @ricardoV94 , thanks for the nudge on the indexing. It's much neater now than using the for loop. Plus I was able to use the full crackers dataset and it fits fast!.

**Comment by NathanielF** [comment]
Giving this one another gentle nudge: @twiecki  and @ricardoV94 

Let me know if there is some concern or anything outstanding?

File: examples/generalized_linear_models/GLM-discrete-choice_models.myst.md [hunk 3]
```diff
@@ -0,0 +1,758 @@
...
+
+To assess overall model adequacy we can rely on the posterior predictive checks to see if the model can recover an approximation to the data generating process.
+
+```{code-cell} ipython3
+fig, axs = plt.subplots(1, 2, figsize=(20, 10))
+ax = axs[0]
+counts = wide_heating_df.groupby("depvar")["idcase"].count()
+predicted_shares = az.extract(idata_m1, var_names=["p"]).mean(axis=2).mean(axis=0)
+ci_lb = np.quantile(az.extract(idata_m1, var_names=["p"]).mean(axis=2), 0.025, axis=0)
+ci_ub = np.quantile(az.extract(idata_m1, var_names=["p"]).mean(axis=2), 0.975, axis=0)
```
Comment (OriolAbril) [inline examples/generalized_linear_models/GLM-discrete-choice_models.myst.md:220]: I would use something similar to the code below, not necessarly the same. As we only want to reduce the chain and draw dimensions, there isn't really need for extract, but it can be kept and after that reduce `sample` instead of both chain and draw as below.

```suggestion
predicted_shares = idata_m1.posterior["p"].mean(dim=["chain", "draw", "obs"])
ci_lb = idata_m1.posterior["p"].mean(dim=["chain", "draw"]).quantile(0.025, dim="obs")
ci_ub = idata_m1.posterior["p"].mean(dim=["chain", "draw"]).quantile(0.975, dim="obs")
```

code using xarray objects should ideally not depend on dimension order but only on dimension _names_.

File: examples/generalized_linear_models/GLM-discrete-choice_models.myst.md [hunk 4]
```diff
@@ -0,0 +1,758 @@
...
+        predictions=True,
+        extend_inferencedata=False,
+        random_seed=100,
+    )
+
+idata_new_policy
+```
+
+```{code-cell} ipython3
+idata_new_policy["predictions"]["p"].stack({"sample": ["chain", "draw"]}).mean(axis=2).mean(axis=0)
```
Comment (OriolAbril) [inline examples/generalized_linear_models/GLM-discrete-choice_models.myst.md:438]:
```suggestion
idata_new_policy["predictions"]["p"].mean(dim=["chain", "draw", "obs"])
```

**Review by OriolAbril — COMMENTED** [review]
added a couple xarray comments, only skimmed the notebook so far but it looks great

**Comment by NathanielF** [comment]
Made those changes @OriolAbril . Thanks for the nudge. Much neater

File: examples/generalized_linear_models/GLM-discrete-choice_models.myst.md [hunk 5]
```diff
@@ -0,0 +1,772 @@
...
+```{code-cell} ipython3
+idata_m1["posterior"]["p"].mean(dim=["chain", "draw", "obs"])
+```
+
+```{code-cell} ipython3
+fig, axs = plt.subplots(1, 2, figsize=(20, 10))
+ax = axs[0]
+counts = wide_heating_df.groupby("depvar")["idcase"].count()
+predicted_shares = idata_m1["posterior"]["p"].mean(dim=["chain", "draw", "obs"])
+ci_lb = idata_m1["posterior"]["p"].quantile(0.025, dim=["chain", "draw", "obs"])
```
Comment (OriolAbril) [inline examples/generalized_linear_models/GLM-discrete-choice_models.myst.md:224]: With the mean it doesn't matter, but for the quantiles, you are now doing a different thing. I am not sure which one you want. Now you take quantiles over all 3 input dims, before you were taking the mean over samples first, then getting the _quantiles of these means_ over the obs dimension.

Comment (NathanielF) [reply examples/generalized_linear_models/GLM-discrete-choice_models.myst.md:224]: Thanks @OriolAbril, yes i think it's the former quantity i want. I want to use it to create the interval on "p" and this way where i take the quantile over the same dimensions as i'm taking the mean puts them on the same scale, which i'm using in the graph to represent the interval.

File: examples/generalized_linear_models/GLM-discrete-choice_models.myst.md [hunk 6]
```diff
@@ -0,0 +1,772 @@
...
+myst:
+  substitutions:
+    extra_dependencies: jax, jaxlib, numpyro
+---
+
+(discrete_choice)=
+# Discrete Choice and Random Utility Models
+
+:::{post} June, 2023
+:tags: categorical regression, generalized linear model, discrete choice
```
Comment (OriolAbril) [inline examples/generalized_linear_models/GLM-discrete-choice_models.myst.md:20]: should we use `model expansion` here too? it is used in the golf putting example. I think it is great to see this kind of examples where we iterate on a model and improve it (it would even be interesting to see an example where we iterate and we often go backwards even but I digress). I think this tag could be used as an indicator of these kind of notebooks: https://pymcio--544.org.readthedocs.build/projects/examples/en/544/blog/tag/model-expansion.html

File: examples/generalized_linear_models/GLM-discrete-choice_models.myst.md [hunk 7]
```diff
@@ -0,0 +1,772 @@
...
+```{code-cell} ipython3
+fig, ax = plt.subplots(figsize=(20, 10))
+
+ax.hist(
+    az.extract(idata_m1["posterior"]["beta_oc"] / idata_m1["posterior"]["beta_ic"])["x"],
+    bins=30,
+    ec="black",
+)
+ax.set_title("Uncertainty in Marginal Rate of Substitution \n Operating Costs / Installation Costs");
+```
```
[the remark this answers is not in the document]
Comment (NathanielF) [reply examples/generalized_linear_models/GLM-discrete-choice_models.myst.md]: Thanks. Good catch. I've adjusted this now.

File: examples/generalized_linear_models/GLM-discrete-choice_models.myst.md [hunk 8]
```diff
@@ -0,0 +1,772 @@
...
+axs[1].set_title("Posterior Predictive Checks")
+ax.set_xlabel("Shares")
+ax.set_ylabel("Heating System");
+```
+
+That extra complexity can be informative, and the degree of relationship amongst the alternative products will inform the substitution patterns under policy changes. Also, note how under this model specification the parameter for `beta_ic` has a expected value of 0. Suggestive perhaps of a resignation towards the reality of installation costs that doesn't change the  utility metric one way or other after a decision to purchase.
+
+```{code-cell} ipython3
+az.summary(
+    idata_m3, var_names=["beta_income", "beta_ic", "beta_oc", "alpha", "chol_corr"], round_to=6
```
[the remark this answers is not in the document]
Comment (NathanielF) [reply examples/generalized_linear_models/GLM-discrete-choice_models.myst.md]: Changed now to 4.

File: examples/generalized_linear_models/GLM-discrete-choice_models.myst.md (hunk 6 above)
Comment (NathanielF) [reply examples/generalized_linear_models/GLM-discrete-choice_models.myst.md]: I've added this tag too.

**Comment by NathanielF** [comment]
Thanks @OriolAbril for the review. I've made those changes now. I've slightly updated and i think improved the final plot but than that I think it should be ready now.

File: examples/generalized_linear_models/GLM-discrete-choice_models.myst.md [hunk 9]
```diff
@@ -0,0 +1,824 @@
...
+    )
+
+    return intervals_df
+
+
+interval_dfs = [get_brand(b) for b in ["nabisco", "keebler", "sunshine"]]
+interval_dfs = pd.concat(interval_dfs, axis=1)
+interval_dfs = interval_dfs.sort_values("means_nabisco")
+interval_dfs.head()
+```
```
[the remark this answers is not in the document]
Comment (OriolAbril) [reply examples/generalized_linear_models/GLM-discrete-choice_models.myst.md:694]: I am not opposed to DataFrames, here the data has already been reduced significantly, so it is not very cumbersome to convert to it. I always try to check the xarray alternative for two reasons.

First and foremost because if we identify tasks that are too difficult to do with xarray objects we can work on fixing that. It also helps me grok xarray (to the point now after some time I can confidently post xarray code I haven't run with some guarantees it will run or nearly run).

Second, because we all know numpy and pandas very well (or most of us quite well at least), so we definitely know how to do X with numpy/pandas, so converting our data to that is often faster/easier to do (even if more verbose) because we know after that we'll be able to do X, whereas if we have to fight with xarray and its docs it might end up taking twice as time. However, if the docs do this, I feel users get the impression that it is not worth it to fight with xarray objects/docs and have no easily accessible examples of how they are useful to them, which is incoherent with the deafult returned object being an InferenceData

**Review by OriolAbril — APPROVED** [review]

File: examples/generalized_linear_models/GLM-discrete-choice_models.myst.md (hunk 9 above)
[the remark this answers is not in the document]
Comment (NathanielF) [reply examples/generalized_linear_models/GLM-discrete-choice_models.myst.md:694]: Makes total sense! 

Thanks again for your time on this one! I really appreciate it.

**Comment by OriolAbril** [comment]
Also, not sure if you are already aware, but while the author metadata isn't very prominently displayed, it is used for aggregating posts by author for example, here is the link to your page: https://www.pymc.io/projects/examples/en/latest/blog/author/nathaniel-forde.html

**Comment by NathanielF** [comment]
> Also, not sure if you are already aware, but while the author metadata isn't very prominently displayed, it is used for aggregating posts by author for example, here is the link to your page: https://www.pymc.io/projects/examples/en/latest/blog/author/nathaniel-forde.html

I did not know that! That's lovely to see!
